### DISCUSSION TITLE
/api/hello.js not able to call from getServerSideProps

### DISCUSSION THREAD

**Discussion by abhikumar22** [vercel/next.js#27584] [Help]
### What version of Next.js are you using?

11.0.0

### What version of Node.js are you using?

v10.22.0

### What browser are you using?

chrome

### What operating system are you using?

macOS

### How are you deploying your application?

Other platform

### Describe the Bug

i have used the default api endpoint which nextjs provides, **/api/hello**. 
in this endpoint i have called my own external backend endpoint (so that it acts as a proxy server to the FE).

I am calling this endpoint from getServerSideProps.

but getting this error :-
`Error: connect ECONNREFUSED 127.0.0.1:80`

When am trying to call the same in client side on button click then its giving result perfectly.

### Expected Behavior

The nextJS endpoint should be called from client as well as from the server

### To Reproduce

1. Create a nextjs API endpoint. -> /api/hello
2. in hello.js call your external api endpoint
3. in any page call **/api/hello** on getServerSideProps.

It will throw the error
`Error: connect ECONNREFUSED 127.0.0.1:80`

**Comment by mikeamarkey** [comment]
Is this related to the problem that you're facing? Problem with calling a next api endpoint from within `getServerSideProps`?
https://github.com/vercel/next.js/discussions/13316

**Comment by abhikumar22** [comment]
yes @mikeamarkey

**Comment by timneutkens** [comment]
https://nextjs.org/docs/basic-features/data-fetching

> Note: You should not use fetch() to call an API route in getServerSideProps. Instead, directly import the logic used inside your API route. You may need to slightly refactor your code for this approach.

> Fetching from an external API is fine!

**Reply by abhikumar22** [reply]
@timneutkens did'nt got much from your reply.

basically you mean to say that we shouldn't call an api route from getServerSideProps. ??

i am trying to make a proxy kind of thing. 
So api route of nextJs will be used as a proxy for both server and client side calls within the application.

In API route in nextjs i will call my actual BE Endpoint.

**Reply by mikeamarkey** [reply]
@abhikumar22 Sorry for not replying more fully, and I know that some time has passed, but for anyone looking for an answer to this question (or related to this question) in the future....

What he's saying is: don't fetch directly in useServerSideProps, instead separate your fetching logic, and import that.

Don't do this:
```
// pages/api/hello
export default async function handler(req, res) {
  const response = await fetch("https://swapi.dev/api/people/1");
  const result = await response.json();
  res.status(200).json(result);
}

// pages/index
export default function Home({ data }) {
  return <div>{JSON.stringify(data)}</div>;
}

export async function getServerSideProps() {
  const data = await fetch('http://localhost/api/hello');

  return {
    props: {
      data: data,
    },
  };
}
```

Do this (or something like this, the key is separating the fetching logic and importing that): 
```
// pages/api/hello
export async function testFetch() {
  const response = await fetch("https://swapi.dev/api/people/1");
  const result = await response.json();
  return result;
}

export default async function handler(req, res) {
  const result = await testFetch();
  res.status(200).json(result);
}

// pages/index
import { testFetch } from "../pages/api/hello";

export default function Home({ data }) {
  return <div>{JSON.stringify(data)}</div>;
}

export async function getServerSideProps() {
  const data = await testFetch();

  return {
    props: {
      data: data,
    },
  };
}

```
